### PR TITLE
[inductor] Kill mark_node_as_mutating

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128979
* __->__ #129346
* #129344
* #129325
* #128893

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -5469,6 +5435,7 @@ def __init__(
         # abi-compatible mode, where we retrieve outputs by pass each individual
         # output through the abi-compatible interface.
         self.outputs: Sequence[Any] = []
+        self.mutation_outputs: List[Buffer] = []
```
Comment (lezcano) [inline torch/_inductor/ir.py:5438]: a bit weird to have an extra attribute just in this class...

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/ir.py]: Fair point. I've moved `mutation_outputs` into `ExternKernel` instead.

**Comment by zou3519** [comment]
Btw this stack broke some internal tests so I might need to revert it. It doesn't revert cleanly right now so I'm looking into it.

**Comment by peterbell10** [comment]
What sort of failures are there?

**Comment by zou3519** [comment]
This was on an internal e2e AOTInductor workflow, so I don't have a minimal repro for you yet. The errors looked like:

1. AssertionError("cuda kernel parameters for triton_per_fused_layer_norm_260 should already exist at this moment, only found dict_keys(['triton_poi_fused_0', 'triton_poi_fused_1', 'triton_poi_fused_2'; which a revert of https://github.com/pytorch/pytorch/pull/129325 fixed

2. tree/torch/_inductor/ir.py", line 3889, in get_read_writes              StarDep(buf.get_name()) for buf in self.get_outputs()                                                                                                                                                                                                      AttributeError: 'ConcatKernel' object has no attribute 'get_outputs'; which a revert of https://github.com/pytorch/pytorch/pull/128893 fixed, but there was a merge conflict so I needed to manually revert.

**Comment by peterbell10** [comment]
> `AttributeError: 'ConcatKernel' object has no attribute 'get_outputs';`

This doesn't make sense as `ConcatKernel` definitely has `get_outputs` defined and it would be completely broken in CI if it wasn't.

**Comment by ColinPeppler** [comment]
@peterbell10 re: Richard's (1)
> AssertionError("cuda kernel parameters for triton_per_fused_layer_norm_260 should already exist at this moment, only found dict_keys(['triton_poi_fused_0', 'triton_poi_fused_1', 'triton_poi_fused_2';

I also ran into the AssertionError with a different model. After a bit of digging around, I saw something weird happening in the code below:

https://github.com/pytorch/pytorch/blob/6dc64026cbadc086a7c2ce3ea9782c33e3928e3b/torch/_inductor/scheduler.py#L1950-L1958

What was weird was on Line 1953: `new_scheduler_node = self.create_scheduler_node(out_buffer)` creates a new scheduler node, but all the new nodes had the same name (e.g. `op2516`).

Is this to be expected?

**Comment by ColinPeppler** [comment]
I did not find a minimal repro but what worked for me was to also replace the operation name here.

https://github.com/pytorch/pytorch/blob/6dc64026cbadc086a7c2ce3ea9782c33e3928e3b/torch/_inductor/scheduler.py#L1918-L1921

**Comment by yf225** [comment]
To avoid op name conflict, should we have a monotonically increasing counter in `V.graph` and use it to name the operation?

**Comment by yf225** [comment]
@peterbell10 I am slightly worried that there will be a long tail of internal test errors that we will need to catch and fix, before we can land this PR stack and keep it in the main branch. Given this, should we continue pushing for landing this stack?

**Comment by peterbell10** [comment]
@ColinPeppler thank you for the reproducer and fix. Should be fixed now by https://github.com/pytorch/pytorch/pull/128893/commits/4f1572656cf6f2e977d43467e4ec7e67cf3359e3

@yf225 I think this is worth doing as we need to support operations with multiple outputs (including mutations) and up till this point this was done by a variety of hacky workarounds. e.g. `MutationOutput` as a separate operation doesn't represent the dependencies correctly and implicitly assumes that the scheduler won't reorder nodes.

**Comment by peterbell10** [comment]
@ColinPeppler would you mind importing this PR to see if the errors are fixed?

**Pull request closed without merging**